rust: Set vlan-filtering 0 on port.vlan: {}

qinqon opened the pull request:
The vlan-filtering at linux-bridge is set to 0 if there is no vlan
configured at ports, but previously with the python version it was also
true if the vlan diciontary was empty. This change implement a fix for
that regression comparing with "default" struct.

qinqon committed:
rust: Set vlan-filtering 0 on port.vlan: {}

The vlan-filtering at linux-bridge is set to 0 if there is no vlan
configured at ports, but previously with the python version it was also
true if the vlan diciontary was empty. This change implement a fix for
that regression comparing with "default" struct.

Signed-off-by: Quique Llorente <ellorent@redhat.com>

cathay4t commented on rust/src/lib/ifaces/linux_bridge.rs:
```diff
@@ -167,7 +167,11 @@ impl LinuxBridgeInterface {
-                port_confs.iter().any(|port_conf| port_conf.vlan.is_some())
+                port_confs.iter().any(|port_conf| {
+                    port_conf.vlan.is_some()
+                        && *port_conf.vlan.as_ref().unwrap()
```
unwrap is not allowed in library

qinqon replied:
Fixed, but how can be the test `test_disable_vlan_filtering_on_bridge` pass without this fix ? 

I have manually test `vlan: {}` on main branch and it's activating vlan_filtering.

qinqon replied:
Well the tests was not testing if vlan-filtering was activated, it's missing something like 

```bash
[ellorent@fedora rust]$ nmcli c show br1 |grep vlan 
bridge.vlan-filtering:                  yes
```

qinqon replied:
Added a check for the vlan-filtering flag

qinqon committed:
test: Check linux bridge vlan-filtering flag

The linux bridge vlan filtering tests are testing just the state but
they are not testing if the vlan-filtering flag is correctly configured.
This change calls npc to check the flag on the test.

Signed-off-by: Quique Llorente <ellorent@redhat.com>

qinqon commented:
Looks like python version is broken
```
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_trunk_port_config[native] PASSED                                                                                                                                      [  9%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_trunk_port_config[not-native] PASSED                                                                                                                                  [ 18%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_port_with_multiple_trunk_ports_config[native] PASSED                                                                                                                  [ 27%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_port_with_multiple_trunk_ports_config[not-native] PASSED                                                                                                              [ 36%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_access_port_config PASSED                                                                                                                                             [ 45%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_update_trunk_port_to_access_port PASSED                                                                                                                               [ 54%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_update_trunk_port_tag_ids PASSED                                                                                                                                      [ 63%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_update_access_port_tag_id PASSED                                                                                                                                      [ 72%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_activate_vlan_filtering_on_bridge PASSED                                                                                                                              [ 81%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_disable_vlan_filtering_on_bridge FAILED                                                                                                                               [ 90%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_pretty_state_port_name_first PASSED                                  
```

While rust is working fine

```
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_trunk_port_config[native] PASSED                                                                                                                                      [  9%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_trunk_port_config[not-native] PASSED                                                                                                                                  [ 18%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_port_with_multiple_trunk_ports_config[native] PASSED                                                                                                                  [ 27%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_port_with_multiple_trunk_ports_config[not-native] PASSED                                                                                                              [ 36%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_access_port_config PASSED                                                                                                                                             [ 45%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_update_trunk_port_to_access_port PASSED                                                                                                                               [ 54%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_update_trunk_port_tag_ids PASSED                                                                                                                                      [ 63%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_update_access_port_tag_id PASSED                                                                                                                                      [ 72%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_activate_vlan_filtering_on_bridge PASSED                                                                                                                              [ 81%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_disable_vlan_filtering_on_bridge PASSED                                                                                                                               [ 90%]
tests/integration/linux_bridge_test.py::TestVlanFiltering::test_pretty_state_port_name_first PASSED                                 
```

Looks like a python regression

qinqon commented:
The testlib make some other test to fail, don't know if they should, since all the differences are with the fields to false, maybe we should just check specific vlan config on the specific tests ? 
```
E       assert False
E        +  where False = <[TypeError("__repr__ returned non-string (type dict)") raised in repr()] method object at 0x7efc4b340a48>(<[TypeError("__repr__ returned non-string (type dict)") raised in repr()] State object at 0x7efc4b28aba8>)
E        +    where <[TypeError("__repr__ returned non-string (type dict)") raised in repr()] method object at 0x7efc4b340a48> = <[TypeError("__repr__ returned non-string (type dict)") raised in repr()] State object at 0x7efc4b28abe0>.match

tests/integration/testlib/assertlib.py:62: AssertionError
----------------------------- Captured stdout call -----------------------------
{'interfaces': [{'name': 'br1', 'type': 'ovs-bridge', 'bridge': {}, 'ipv4': {'enabled': False, 'address': [], 'dhcp': False}, 'ipv6': {'enabled': False, 'address': [], 'dhcp': False, 'autoconf': False}, 'lldp': {'enabled': False}}]}
{'interfaces': [{'name': 'br1', 'type': 'ovs-bridge', 'ipv4': {'enabled': False, 'address': [], 'dhcp': False}, 'ipv6': {'enabled': False, 'address': [], 'dhcp': False, 'autoconf': False}, 'lldp': {'enabled': False}, 'bridge': {'options': {'stp': False, 'rstp': False, 'mcast-snooping-enable': False, 'fail-mode': ''}, 'port': []}}]}
```

Clearly is failing because the state does not match, should it match ?

qinqon commented:
@cathay4t the main problem though is that the python nmstate version is not reseting port's vlan filtering if `vlan: {}` I am pretty sure that is working at the nmstate version cosumed at kubernetes-nmstate since we have a e2e test for it.

qinqon committed:
test: Use assert_state at vlan disabling test

The test that check that is possible to disable vlan-filtering using
`vlan: {}` need to assert state with `assert_state` so it will not skip
empty vlan struct.

Signed-off-by: Quique Llorente <ellorent@redhat.com>

qinqon commented:
Some tests are failing on 
```python
        for iface_state in desired_state[Interface.KEY]:
>           if iface_state[Interface.IPV4][InterfaceIPv4.ENABLED]:
E           KeyError: 'enabled'
```

ffmancera commented on libnmstate/state.py:
```diff
@@ -101,6 +101,9 @@ def merge_dict(dict_to, dict_from):
     """
+    # If was explicitly set empty it has to reset that part of the state.
+    if len(dict_to) == 0:
+        return
```
This is probably breaking ipv4 when not specified or empty. Because for IPv4 we rely on NetworkManager. Let me think on a better solution. Will send you a patch before the end of the day :-)

ffmancera committed:
python, bridge: clear vlan filtering when set empty

When the user specifies 'vlan: {}', Nmstate should clear the vlan
configuration for that port instead of merging it.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

cathay4t commented on tests/integration/linux_bridge_test.py:
```diff
@@ -147,6 +147,21 @@ def bond0(port0_up):
+        ),
+        check=True,
+    )
+    return vlan_filtering_flag.strip() == "yes"
```
Instead of checking `nmcli` output, please try use `npc iface br0 -j` or `ip  -d  link show br0` which are  checking the kernel state instead of on-disk configurations.

ffmancera replied:
Done, thanks!

cathay4t reviewed (changes requested):
WIth your PR, this yaml also disable vlan filtering:

```yml
---
interfaces:
  - name: br0
    type: linux-bridge
    state: up
    bridge:
      port:
        - name: eth1
        - name: eth2
```

From the commit message, you only disable vlan filtering with:

```yml
    bridge:
      port:
        - name: eth1
          vlan: {}
        - name: eth2
          vlan: {}
```

cathay4t commented:
~~I am OK to override VLAN fitlering information whenever port list defeind, but we need to mention that in commit message and test it in CI.
Rust unit test cases could be a good choice to reveal that.~~

ffmancera commented:
> WIth your PR, this yaml also disable vlan filtering:
> 
> ```yaml
> ---
> interfaces:
>   - name: br0
>     type: linux-bridge
>     state: up
>     bridge:
>       port:
>         - name: eth1
>         - name: eth2
> ```
> 
> From the commit message, you only disable vlan filtering with:
> 
> ```yaml
>     bridge:
>       port:
>         - name: eth1
>           vlan: {}
>         - name: eth2
>           vlan: {}
> ```

I will check this, if that is the case, it must be improved and tested.

ffmancera commented:
> > WIth your PR, this yaml also disable vlan filtering:
> > ```yaml
> > ---
> > interfaces:
> >   - name: br0
> >     type: linux-bridge
> >     state: up
> >     bridge:
> >       port:
> >         - name: eth1
> >         - name: eth2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > From the commit message, you only disable vlan filtering with:
> > ```yaml
> >     bridge:
> >       port:
> >         - name: eth1
> >           vlan: {}
> >         - name: eth2
> >           vlan: {}
> > ```
> 
> I will check this, if that is the case, it must be improved and tested.

As I said, I tried this out and it is working fine for me:

Given the following state applied
```
interfaces:
- name: linux-br0
  type: linux-bridge
  state: up
  accept-all-mac-addresses: false
  bridge:
    options:
      gc-timer: 9306
      group-addr: 01:80:C2:00:00:00
      group-forward-mask: 0
      hash-max: 4096
      hello-timer: 110
      mac-ageing-time: 300
      multicast-last-member-count: 2
      multicast-last-member-interval: 100
      multicast-querier: false
      multicast-querier-interval: 25500
      multicast-query-interval: 12500
      multicast-query-response-interval: 1000
      multicast-query-use-ifaddr: false
      multicast-router: 1
      multicast-snooping: true
      multicast-startup-query-count: 2
      multicast-startup-query-interval: 3125
      stp:
        enabled: true
        forward-delay: 15
        hello-time: 2
        max-age: 20
        priority: 32768
    port:
    - name: dummy1
      stp-hairpin-mode: false
      stp-path-cost: 100
      stp-priority: 32
      vlan:
        enable-native: true
        mode: trunk
        tag: 100
        trunk-tags:
        - id: 101
        - id-range:
            max: 599
            min: 500
```

If I apply the following state:

```
interfaces:
  - name: dummy1
    type: dummy
    state: up
  - name: linux-br0
    type: linux-bridge
    state: up
    bridge:
      port:
        - name: dummy1
          stp-hairpin-mode: false
          stp-path-cost: 100
          stp-priority: 32
```

then the result will contain the vlan configuration see:

```
root@8a654ab8889d nmstate-workspace]# nmstatectl show linux-br0
2022-03-25 13:52:04,873 root         DEBUG    NetworkManager version 1.33.4
2022-03-25 13:52:04,874 root         DEBUG    Async action: Retrieve applied config: dummy dummy1 started
2022-03-25 13:52:04,874 root         DEBUG    Async action: Retrieve applied config: bridge linux-br0 started
2022-03-25 13:52:04,875 root         DEBUG    Async action: Retrieve applied config: dummy dummy1 finished
2022-03-25 13:52:04,875 root         DEBUG    Async action: Retrieve applied config: bridge linux-br0 finished
2022-03-25 13:52:04,876 root         DEBUG    Interface dummy.dummy1 found. Merging the interface information.
2022-03-25 13:52:04,877 root         DEBUG    Interface ethernet.eth1 found. Merging the interface information.
2022-03-25 13:52:04,877 root         DEBUG    Interface ethernet.eth2 found. Merging the interface information.
2022-03-25 13:52:04,877 root         DEBUG    Interface linux-bridge.linux-br0 found. Merging the interface information.
---
dns-resolver:
  config: {}
  running: {}
route-rules:
  config: []
routes:
  config: []
  running: []
interfaces:
- name: linux-br0
  type: linux-bridge
  state: up
  accept-all-mac-addresses: false
  bridge:
    options:
      gc-timer: 2085
      group-addr: 01:80:C2:00:00:00
      group-forward-mask: 0
      hash-max: 4096
      hello-timer: 89
      mac-ageing-time: 300
      multicast-last-member-count: 2
      multicast-last-member-interval: 100
      multicast-querier: false
      multicast-querier-interval: 25500
      multicast-query-interval: 12500
      multicast-query-response-interval: 1000
      multicast-query-use-ifaddr: false
      multicast-router: 1
      multicast-snooping: true
      multicast-startup-query-count: 2
      multicast-startup-query-interval: 3125
      stp:
        enabled: true
        forward-delay: 15
        hello-time: 2
        max-age: 20
        priority: 32768
    port:
    - name: dummy1
      stp-hairpin-mode: false
      stp-path-cost: 100
      stp-priority: 32
      vlan:
        enable-native: true
        mode: trunk
        tag: 100
        trunk-tags:
        - id: 101
        - id-range:
            max: 599
            min: 500
  ethtool:
    feature:
      highdma: true
      rx-gro: true
      rx-gro-list: false
      rx-udp-gro-forwarding: false
      tx-checksum-ip-generic: true
      tx-esp-segmentation: true
      tx-fcoe-segmentation: false
      tx-generic-segmentation: true
      tx-gre-csum-segmentation: true
      tx-gre-segmentation: true
      tx-gso-list: true
      tx-gso-partial: true
      tx-gso-robust: false
      tx-ipxip4-segmentation: true
      tx-ipxip6-segmentation: true
      tx-nocache-copy: false
      tx-scatter-gather-fraglist: true
      tx-sctp-segmentation: true
      tx-tcp-ecn-segmentation: true
      tx-tcp-mangleid-segmentation: true
      tx-tcp-segmentation: true
      tx-tcp6-segmentation: true
      tx-tunnel-remcsum-segmentation: true
      tx-udp-segmentation: true
      tx-udp_tnl-csum-segmentation: true
      tx-udp_tnl-segmentation: true
      tx-vlan-hw-insert: true
      tx-vlan-stag-hw-insert: true
  ipv4:
    enabled: false
    address: []
    dhcp: false
  ipv6:
    enabled: false
    address: []
    autoconf: false
    dhcp: false
  lldp:
    enabled: false
  mac-address: 5E:9D:E8:B8:4E:44
  mtu: 1500
ovs-db:
  external_ids:
    hostname: 8a654ab8889d
    rundir: /var/run/openvswitch
    system-id: 4f0afa4e-6adb-4aa0-b883-ef5230e4961d
  other_config: {}

```
I will address the other comments, what do you think?

cathay4t commented:
The rust code is causing vlan filtering been disabled with

```yml
    bridge:
      port:
        - name: eth1
        - name: eth2
```

And the same yml file been treated as preserving existing VLAN filtering configure when using python code.

So please fix this PR.

qinqon commented:
> > > WIth your PR, this yaml also disable vlan filtering:
> > > ```yaml
> > > ---
> > > interfaces:
> > >   - name: br0
> > >     type: linux-bridge
> > >     state: up
> > >     bridge:
> > >       port:
> > >         - name: eth1
> > >         - name: eth2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > From the commit message, you only disable vlan filtering with:
> > > ```yaml
> > >     bridge:
> > >       port:
> > >         - name: eth1
> > >           vlan: {}
> > >         - name: eth2
> > >           vlan: {}
> > > ```
> > 
> > 
> > I will check this, if that is the case, it must be improved and tested.
> 
> As I said, I tried this out and it is working fine for me:
> 
> Given the following state applied
> 
> ```
> interfaces:
> - name: linux-br0
>   type: linux-bridge
>   state: up
>   accept-all-mac-addresses: false
>   bridge:
>     options:
>       gc-timer: 9306
>       group-addr: 01:80:C2:00:00:00
>       group-forward-mask: 0
>       hash-max: 4096
>       hello-timer: 110
>       mac-ageing-time: 300
>       multicast-last-member-count: 2
>       multicast-last-member-interval: 100
>       multicast-querier: false
>       multicast-querier-interval: 25500
>       multicast-query-interval: 12500
>       multicast-query-response-interval: 1000
>       multicast-query-use-ifaddr: false
>       multicast-router: 1
>       multicast-snooping: true
>       multicast-startup-query-count: 2
>       multicast-startup-query-interval: 3125
>       stp:
>         enabled: true
>         forward-delay: 15
>         hello-time: 2
>         max-age: 20
>         priority: 32768
>     port:
>     - name: dummy1
>       stp-hairpin-mode: false
>       stp-path-cost: 100
>       stp-priority: 32
>       vlan:
>         enable-native: true
>         mode: trunk
>         tag: 100
>         trunk-tags:
>         - id: 101
>         - id-range:
>             max: 599
>             min: 500
> ```
> 
> If I apply the following state:
> 
> ```
> interfaces:
>   - name: dummy1
>     type: dummy
>     state: up
>   - name: linux-br0
>     type: linux-bridge
>     state: up
>     bridge:
>       port:
>         - name: dummy1
>           stp-hairpin-mode: false
>           stp-path-cost: 100
>           stp-priority: 32
> ```
> 
> then the result will contain the vlan configuration see:
> 
> ```
> root@8a654ab8889d nmstate-workspace]# nmstatectl show linux-br0
> 2022-03-25 13:52:04,873 root         DEBUG    NetworkManager version 1.33.4
> 2022-03-25 13:52:04,874 root         DEBUG    Async action: Retrieve applied config: dummy dummy1 started
> 2022-03-25 13:52:04,874 root         DEBUG    Async action: Retrieve applied config: bridge linux-br0 started
> 2022-03-25 13:52:04,875 root         DEBUG    Async action: Retrieve applied config: dummy dummy1 finished
> 2022-03-25 13:52:04,875 root         DEBUG    Async action: Retrieve applied config: bridge linux-br0 finished
> 2022-03-25 13:52:04,876 root         DEBUG    Interface dummy.dummy1 found. Merging the interface information.
> 2022-03-25 13:52:04,877 root         DEBUG    Interface ethernet.eth1 found. Merging the interface information.
> 2022-03-25 13:52:04,877 root         DEBUG    Interface ethernet.eth2 found. Merging the interface information.
> 2022-03-25 13:52:04,877 root         DEBUG    Interface linux-bridge.linux-br0 found. Merging the interface information.
> ---
> dns-resolver:
>   config: {}
>   running: {}
> route-rules:
>   config: []
> routes:
>   config: []
>   running: []
> interfaces:
> - name: linux-br0
>   type: linux-bridge
>   state: up
>   accept-all-mac-addresses: false
>   bridge:
>     options:
>       gc-timer: 2085
>       group-addr: 01:80:C2:00:00:00
>       group-forward-mask: 0
>       hash-max: 4096
>       hello-timer: 89
>       mac-ageing-time: 300
>       multicast-last-member-count: 2
>       multicast-last-member-interval: 100
>       multicast-querier: false
>       multicast-querier-interval: 25500
>       multicast-query-interval: 12500
>       multicast-query-response-interval: 1000
>       multicast-query-use-ifaddr: false
>       multicast-router: 1
>       multicast-snooping: true
>       multicast-startup-query-count: 2
>       multicast-startup-query-interval: 3125
>       stp:
>         enabled: true
>         forward-delay: 15
>         hello-time: 2
>         max-age: 20
>         priority: 32768
>     port:
>     - name: dummy1
>       stp-hairpin-mode: false
>       stp-path-cost: 100
>       stp-priority: 32
>       vlan:
>         enable-native: true
>         mode: trunk
>         tag: 100
>         trunk-tags:
>         - id: 101
>         - id-range:
>             max: 599
>             min: 500
>   ethtool:
>     feature:
>       highdma: true
>       rx-gro: true
>       rx-gro-list: false
>       rx-udp-gro-forwarding: false
>       tx-checksum-ip-generic: true
>       tx-esp-segmentation: true
>       tx-fcoe-segmentation: false
>       tx-generic-segmentation: true
>       tx-gre-csum-segmentation: true
>       tx-gre-segmentation: true
>       tx-gso-list: true
>       tx-gso-partial: true
>       tx-gso-robust: false
>       tx-ipxip4-segmentation: true
>       tx-ipxip6-segmentation: true
>       tx-nocache-copy: false
>       tx-scatter-gather-fraglist: true
>       tx-sctp-segmentation: true
>       tx-tcp-ecn-segmentation: true
>       tx-tcp-mangleid-segmentation: true
>       tx-tcp-segmentation: true
>       tx-tcp6-segmentation: true
>       tx-tunnel-remcsum-segmentation: true
>       tx-udp-segmentation: true
>       tx-udp_tnl-csum-segmentation: true
>       tx-udp_tnl-segmentation: true
>       tx-vlan-hw-insert: true
>       tx-vlan-stag-hw-insert: true
>   ipv4:
>     enabled: false
>     address: []
>     dhcp: false
>   ipv6:
>     enabled: false
>     address: []
>     autoconf: false
>     dhcp: false
>   lldp:
>     enabled: false
>   mac-address: 5E:9D:E8:B8:4E:44
>   mtu: 1500
> ovs-db:
>   external_ids:
>     hostname: 8a654ab8889d
>     rundir: /var/run/openvswitch
>     system-id: 4f0afa4e-6adb-4aa0-b883-ef5230e4961d
>   other_config: {}
> ```
> 
> I will address the other comments, what do you think?

@ffmancera that's fine by me, about the scenario I checked and they were not removing the vlan config when is not specified.

ffmancera committed:
rust, linux-bridge: keep vlan-filtering when not set

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

qinqon commented:
@ffmancera everything is fine we can merge it.
